contracts/LightClient.sol

```diff
@@ -144,11 +154,62 @@ contract LightClient is ILightClient, Ownable {
         require(_epochSigners.length == _epochVotingPowers.length, "incorrect length");
 
         uint256 position = _epochPosition(_epochIdx);
+
         curEpochIdx = _epochIdx;
         epochs[position].curEpochVals = _epochSigners;
         epochs[position].curVotingPowers = _epochVotingPowers;
+    }
+
+    function _totalProduceBlock(uint256[] memory produceBlocks) internal pure returns (uint256 total) {
+        for (uint256 i = 0; i < produceBlocks.length; i++) {
+            total += produceBlocks[i];
+        }
+    }
+
+    function _validatorRewardShare(
+        uint256 epochReward,
+        uint256 produceAmount,
+        uint256 totalProduceAmount
+    ) internal pure returns (uint256) {
+        return (epochReward * produceAmount) / totalProduceAmount;
+    }
+
+    function _perEpochReward(address[] memory rewardVals, uint256[] memory produceAmountList) internal {
+        uint256 epochReward = w3qErc20.perEpochReward();
+
+        // Calculate the amount of tokens to reward validator and delegators
+        for (uint256 i = 0; i < rewardVals.length; i++) {
+            address valAddr = rewardVals[i];
 
-        // TODO: add rewards to validators
+            uint256 valShares = staking.getValidatorShare(valAddr);
+            address[] memory delAddrs = staking.getDelegatorAddrs(valAddr);
+
+            uint256 totalRewardAmount = _validatorRewardShare(
+                epochReward,
+                produceAmountList[i],
+                _totalProduceBlock(produceAmountList)
+            );
+            uint256 valRewardAmount = totalRewardAmount;
+
+            for (uint256 j = 0; j < delAddrs.length; j++) {
```

Comment:
I guess this will cause a huge gas with a long delegator list. A simple way may add  `rewardValidator(val, amount)` in the Staking contract, which just simply transferFrom(token), and then adds tokens to the validator balance.